macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -38,113 +35,156 @@ def expo_weights_arr(lback_periods: int, half_life: int, *args, **kwargs) -> np.
-        The function must have a signature of (np.ndarray, *, remove_zeros: bool) -> float.
-        See `expo_std` and `flat_std` for examples.
-    :param <bool> remove_zeros: removes zeroes as invalid entries and shortens the
-        effective window.
-    :param <np.ndarray> weights: array of exponential weights for each period in the
-        lookback window. Default is None, which means that the weights are equal.
     """
-    if weights is None:
-        weights = np.ones(df_wide.shape[0]) / df_wide.shape[0]
-        univariate_vol: pd.Series = df_wide.agg(roll_func, remove_zeros=remove_zeros)
-    else:
-        if len(weights) == len(df_wide):
-            univariate_vol = df_wide.agg(
-                roll_func, w=weights, remove_zeros=remove_zeros
-            )
-        else:
-            return pd.Series(np.nan, index=df_wide.columns)
-
-        ## Creating a mask to fill series `nan_tolerance`
-    mask = (
-        (
-            df_wide.isna().sum(axis=0)
-            + (df_wide == 0).sum(axis=0)
-            + (lback_periods - len(df_wide))
-        )
-        / lback_periods
-    ) <= nan_tolerance
+    assert x.ndim == 1 or x.shape[1] == 1, "`x` must be a 1D array or a column vector"
+    assert y.ndim == 1 or y.shape[1] == 1, "`y` must be a 1D array or a column vector"
+    assert x.shape[0] == y.shape[0], "`x` and `y` must have same length"
+    assert w.ndim == 1 or w.shape[1] == 1, "`w` must be a 1D array or a column vector"
+    assert x.shape[0] == w.shape[0], "`x` and `w` must have same length"
+
+    w = w / sum(w)  # weights are normalized
 
-    univariate_vol[~mask] = np.nan
+    rss = (x - x.mean()) * (y - y.mean())
 
-    return univariate_vol
+    return w.T.dot(rss)
 
 
-def weighted_covariance(
-    x: np.ndarray, y: np.ndarray, w: np.ndarray = None, remove_zeros: bool = True
-):
-    """
-    Estimate covariance between two series after applying weights.
+def _nan_ratio(x, remove_zeros: bool = True) -> float:
+    return (sum(np.isnan(x)) + (sum(x == 0) if remove_zeros else 0)) / len(x)
 
-    """
-    # assert len(x) == len(w), "weights and window must have same length"
+
+def _mask_nan_series(
+    x: np.ndarray,
+    y: np.ndarray,
+    lback_periods: int,
+    remove_zeros: bool,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Mask NaNs in the series."""
+    # if either of the dates is NaN, mask that row
+    mask = np.isnan(x) | np.isnan(y)
     if remove_zeros:
-        x = x[x != 0]
-        w = w[0 : len(x)] / sum(w[0 : len(x)])
-    w = w / sum(w)  # weights are normalized
+        mask = mask | (x == 0) | (y == 0)
 
-    if y is not None:
-        rss = (x - x.mean()) * (y - y.mean())
-    else:
-        rss = (x - x.mean()) ** 2
+    # drop NaNs
+    x = x[~mask]
+    y = y[~mask]
 
-    return w.T.dot(rss)
+    # if dropping nans leaves less than lback_periods, fill with NaNs and return
+    if any(len(_) < lback_periods for _ in [x, y]):
```

Comment:
@Magnus167 in my books the use of `_` is a misnomer. Unless I am mistaken standard practise it to use `_` for unused variables?

Reply:
It's used for a temporary variable; typically used when you want need a variable to pass/store/show a value, but don't need the variable afterwards.

Reply:
No mistake, @lsimonsen.
The underscore signifies that the variable should be ignored because it is not used in the loop.
It's also sometimes used in cases where a function might return multiple values, but some are to be ignored:
```
x, _ = func()
```

Reply:
I think I got it - the above `x, _ = func()` was what I was expecting (it never gets used). However, the point from you two, is that `_` can also be used inside the conditional statement `if statement`, when it doesn't get used afterwards/outside that evaluation.

Reply:
Thank you both @emhbrine and @Magnus167